fix(createAsyncStoragePersister): persistClient not respect throttleTime and not persisted final state queryClient (#3331)

huanguolin opened the pull request:
closes #3331

fix(createAsyncStoragePersister): persistClient not respect throttleTime and not persisted final state queryClient (#3331).
And add unit test for asyncThrottle.

TkDodo reviewed (approved):
thank you 🙏

TkDodo commented on src/createAsyncStoragePersister/asyncThrottle.ts:
```diff
@@ -0,0 +1,51 @@
+      try {
+        running = true
+        await func(...args)
+      } catch (error) {
+        getLogger().error(error)
+      } finally {
```
one small thing: in v4, we put all loggings behind a production env check, like here:

https://github.com/tannerlinsley/react-query/blob/fe98d533cc62fab8ab88166a31707d6cfc02d058/src/core/query.ts#L449-L451

can we do that here as well please

TkDodo replied:
oh but actually, I don't think we should be logging here on this level. We have [another PR](https://github.com/tannerlinsley/react-query/pull/3246) where we removed the global logger and made it an option on the queryClient instead, so this would be problematic.

If we want to catch errors and log them, how about doing it in the `persistQueryClient` itself, when the writing happens? I think this would be somewhere around here:

https://github.com/tannerlinsley/react-query/blob/1abb3714637c3f9044a4294cc737cc41966a2f49/src/persistQueryClient/index.ts#L114

Not sure if it makes sense to just swallow and log the error though. Shouldn't we at least re-throw it to the consumers?

huanguolin commented:
@TkDodo 
I totally agree with you.
The asyncThrottle has to catch errors to avoid state corruption (like running). I currently have asyncThrottle provide an onError callback.
It's right to handle errors in `persistQueryClient`, but I'm not sure how it's appropriate to do that at the moment: 

- Is it to use a user-uniformly configured error handling callback from `queryClient`?
- Or do I modify the createAsyncStoragePersister interface?
- Or split this to another pr, to be handled in the future?

TkDodo commented:
We can do another PR and think about it. I don't think errors were handled in the previous version either

huanguolin commented:
@TkDodo 
OK!
The current pull request has been updated, we can now move on...

TkDodo commented on src/createAsyncStoragePersister/asyncThrottle.ts:
```diff
@@ -0,0 +1,56 @@
+      try {
+        running = true
+        await func(...args)
+      } catch (error) {
+        onError(error)
+      } finally {
```
hm, I don't fully understand why we would need the `onError` callback. If remove the `catch` statement here and just let the error bubble up to the call side, what would happen?

huanguolin replied:
Remove `catch` directly, keep `try...finally` is fine for `asyncThrottle`. But then error handling must be handled internally by the `func` itself. `asyncThrottle` returns a non-waitable function.

If we decide to handle errors inside the `func`, I'll change it right away.

TkDodo replied:
i think it's expected that if `asyncThrottle` returns a `Promise` that it can fail

huanguolin replied:
Sorry, I wasn't clear.

If the `onError` callback is retained, the error handling will look like this：
```typescript
return {
    persistClient: asyncThrottle(
      (persistedClient: PersistedClient) => storage.setItem(key, serialize(persistedClient)),
      { 
        interval: throttleTime,
        onError: (err: unknown) => {
          // handing error here
        } 
      }
    ),
// ....
```

If the `onError` callback is removed directly：
```typescript
return {
    persistClient: asyncThrottle(
      async (persistedClient: PersistedClient) => {
        try {
          storage.setItem(key, serialize(persistedClient))
        } catch (error) {
          // handing error here
        }
      }, 
      throttleTime
    ),
// ....
```



> i think it's expected that if `asyncThrottle` returns a `Promise` that it can fail

`asyncThrottle` returns a non-waitable function. Do you mean to return a function that returns a `promise`?
If so, this is strange. `asyncThrottle` returns a function that may be called 10 times， but the 'func' may be executed only twice. `promise` either resolves or rejects, and cannot be left dangling.
And the code will be more complex than now, I still think the previous two better a little , do you think?
☺

TkDodo replied:
I see, my mistake. let's keep the callback then 👍

TkDodo commented:
getting a lint error now:

```
/home/runner/work/react-query/react-query/src/createAsyncStoragePersister/asyncThrottle.ts
Error:   6:20  error  Unexpected empty arrow function  @typescript-eslint/no-empty-function
```

huanguolin commented:
Fixed. And improve unit test coverage.

TkDodo commented:
prettier fails now:

```
$ prettier "{.,src,src/**,example/src,example/src/**,types}/*.{md,js,jsx,ts,tsx,json}" --check
Checking formatting...
[warn] src/createAsyncStoragePersister/asyncThrottle.ts
[warn] src/createAsyncStoragePersister/tests/asyncThrottle.test.ts
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
```

huanguolin commented:
@TkDodo 
I'm sorry for that.
I had problems with my local environment, I changed a lot of version numbers in package.json, and the configuration of `.eslintrc`, and I was barely able to run eslint. But I was still prompted with some errors, but not caused by my commit this time (I didn't commit the code just to make eslint work). 
![image](https://user-images.githubusercontent.com/12430413/156138052-0f4a6fe9-8f5d-4483-b889-5c25ec74aa54.png)

I've fixed the formatting problem. Hope everything goes well this time. No more wasting your time.

tannerlinsley commented:
:tada: This PR is included in version 4.0.0-alpha.19 :tada:

The release is available on:
- [npm package (@alpha dist-tag)](https://www.npmjs.com/package/react-query/v/4.0.0-alpha.19)
- [GitHub release](https://github.com/tannerlinsley/react-query/releases/tag/v4.0.0-alpha.19)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tannerlinsley commented:
:tada: This PR is included in version 4.0.0-beta.1 :tada:

The release is available on:
- [npm package (@beta dist-tag)](https://www.npmjs.com/package/react-query/v/4.0.0-beta.1)
- [GitHub release](https://github.com/tannerlinsley/react-query/releases/tag/v4.0.0-beta.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
